### ISSUE TITLE
[R-Forge #5135] Avoid "eval" in Cdogroups wherever possible

### ISSUE THREAD

**Issue by arunsrinivasan** [Rdatatable/data.table#352]
Submitted by: Arun ; Assigned to: Nobody; [R-Forge link](https://r-forge.r-project.org/tracker/index.php?func=detail&aid=5135&group_id=240&atid=978)

The "eval" call in Cdogroups is very nice when the number of groups are less. But when there are over 1e7 groups, for example, even a primitive function's call (that many times) becomes quite expensive. Here's an example:

DT <- data.table(x=1:1e7, y=1L)
setattr(DT, 'sorted', 'x')
system.time(ans <- DT[, sum(y), by=x])
   user  system elapsed 
 12.556   0.722  14.901 

First, note that a for-loop with 1e7 calls to "sum" takes 7 seconds. 

system.time(for(i in 1:1e7) sum(1L))
   user  system elapsed 
  6.355   0.043   7.061 

Next, just adding an "eval" makes it a lot worse:

expr <- quote(sum(1L))
system.time(for(i in 1:1e7) eval(expr))
   user  system elapsed 
 38.920   0.223  40.663 

Basically, if we can implement/set up certain commonly used functions like "sum", "mean", "var", "sd" etc.. in data.table and also detect and call from within C directly, then we can save _a LOT of time_!
